Kuzzle/Offline/Query/QueryReplayer.cs

```diff
@@ -0,0 +1,225 @@
+    void RejectQueries(Predicate<JObject> predicate, Exception exception);
+    int Remove(Predicate<JObject> predicate);
+
+    CancellationTokenSource ReplayQueries(bool resetWaitLogin = true);
+    CancellationTokenSource ReplayQueries(Predicate<JObject> predicate, bool resetWaitLogin = true);
+
+    void CancelReplay();
+  }
+
+  internal sealed class QueryReplayer : IQueryReplayer {
+    private IKuzzle kuzzle;
+    private List<TimedQuery> queue;
+    private IOfflineManager offlineManager;
+    private CancellationTokenSource cancellationTokenSource;
+    private bool currentlyReplaying = false;
+    private Stopwatch stopWatch = new Stopwatch();
+
+    /// <summary>
+    /// Tells if the QueryReplayer is locked (i.e. it doesn't accept new queries).
+    /// Should only be true if there is an 'auth:login' or 'auth:logout' call in the queue
+    /// </summary>
+    public bool Lock { get; set; } = false;
+
+    /// <summary>
+    /// True if the QueryReplayer has to wait a login call before replaying the queue
+    /// </summary>
+    public bool WaitLoginToReplay { get; set; } = false;
+
+    /// <summary>
+    /// Constructor of the QueryReplayer class.
+    /// </summary>
+    internal QueryReplayer(IOfflineManager offlineManager, IKuzzle kuzzle) {
+      this.queue = new List<TimedQuery>();
+      this.offlineManager = offlineManager;
+      this.kuzzle = kuzzle;
+      this.cancellationTokenSource = new CancellationTokenSource();
+      this.ReplayQuery = ReplayOneQuery;
+    }
+
+    /// <summary>
+    /// Add a new query to the queue.
+    /// Return true if successful
+    /// </summary>
+    public bool Enqueue(JObject query) {
+      if (Lock) return false;
+
+      lock (queue) {
+        if (queue.Count < offlineManager.MaxQueueSize || offlineManager.MaxQueueSize < 0) {
+          if (queue.Count == 0) {
+            stopWatch.Reset();
+            stopWatch.Start();
+            queue.Add(new TimedQuery(query, 0));
+          } else {
+            TimedQuery previous = queue[queue.Count - 1];
+            Int64 elapsedTime = stopWatch.ElapsedMilliseconds - previous.Time;
+            elapsedTime = Math.Min(elapsedTime, offlineManager.MaxRequestDelay);
+            queue.Add(new TimedQuery(query, previous.Time + elapsedTime));
+          }
+          if (query["controller"]?.ToString() == "auth"
+              && (query["action"]?.ToString() == "login"
+                || query["action"]?.ToString() == "logout")
+              ) {
+                Lock = true;
+              }
+
+          return true;
+        }
+      }
+      return false;
+    }
+
+    public int Count {
+      get { return queue.Count; }
+    }
+
+    /// <summary>
+    /// Remove and return the first query that has been added to the queue.
+    /// </summary>
+    public JObject Dequeue() {
+      lock (queue) {
+        if (queue.Count == 0)
+          throw new InvalidOperationException("Cannot dequeue an Empty queue");
```

Comment:
Why is it invalid to dequeue an empty queue? Usually that kind of operation results in a noop.